src/script_zones.c

```diff
@@ -0,0 +1,289 @@
+
+#include "config_creature.h"
+#include "creature_states.h"
+#include "frontmenu_ingame_map.h"
+#include "game_merge.h"
+#include "slab_data.h"
+#include "map_data.h"
+#include "map_columns.h"
+#include "room_data.h"
+#include "room_jobs.h"
+#include "room_util.h"
+#include "thing_objects.h"
+
+struct ScriptZoneRecord *add_script_zone()
+{
+    return &gameadd.script_zone_records[++gameadd.max_valid_script_zone_recods];
+}
+
+int script_zone_id(struct ScriptZoneRecord *zone)
+{
+    return zone - gameadd.script_zone_records;
+}
+
+struct ScriptZoneRecord *get_script_zone(int zone_id)
+{
+    if ((zone_id < 0) || (zone_id >= MAX_ZONE_RECORDS))
+    {
+        return INVALID_SCRIPT_ZONE;
+    }
+    return &gameadd.script_zone_records[zone_id];
+}
+
+struct ScriptZoneRecord *find_script_zone(int zone_id)
+{
+    if (zone_id <= 0)
+    {
+        return INVALID_SCRIPT_ZONE;
+    }
+    for (int i = 1; i <= gameadd.max_valid_script_zone_recods; i++)
+    {
+        struct ScriptZoneRecord *ptr = &gameadd.script_zone_records[i];
+        if (ptr->zone_id == zone_id)
+        {
+            return ptr;
+        }
+    }
+    return INVALID_SCRIPT_ZONE;
+}
+
+static void swap_zones(struct ScriptZoneRecord *first, struct ScriptZoneRecord *second)
+{
+    for (int dy = 0; dy < first->hheight * STL_PER_SLB; dy++)
+    {
+        for (int dx = 0; dx < first->hwidth * STL_PER_SLB; dx++)
+        {
+            struct Map *sblk = get_map_block_at(slab_subtile(first->min_x, 0) + dx, slab_subtile(first->min_y, 0) + dy);
+            struct Map *dblk = get_map_block_at(slab_subtile(second->min_x, 0) + dx, slab_subtile(second->min_y, 0) + dy);
+            // moving columns and visibility
+            struct Map old_blk = *dblk;
+            *dblk = *sblk;
+            *sblk = old_blk;
+
+            struct Thing *tng;
+            // moving things on dblk
+            for (int idx = get_mapwho_thing_index(dblk);
+                idx != 0;
+                )
+            {
+                tng = thing_get(idx);
+                idx = tng->next_on_mapblk;
+                // Remove all creatures from rooms. We will add them back later
+                struct CreatureControl *cctrl = creature_control_get_from_thing(tng);
+                if (!creature_control_invalid(cctrl))
+                {
+                    remove_creature_from_work_room(tng);
+                }
+                // We should adjust mappos for deleted things because deleting would query location
+                tng->mappos.x.stl.num -= slab_subtile(first->min_x, 0);
+                tng->mappos.x.stl.num += slab_subtile(second->min_x, 0);
+                tng->mappos.y.stl.num -= slab_subtile(first->min_y, 0);
+                tng->mappos.y.stl.num += slab_subtile(second->min_y, 0);
+                if (tng->class_id == TCls_Object)
+                {
+                    switch (tng->model)
+                    {
+                        // Room flames
+                        case 55:
```

Comment:
What are these objects?

Reply:
Different room flames (these sparks circling around room). They would be recreated on room reintegration.